src/nvim/terminal.c

```diff
@@ -465,12 +453,8 @@ static int terminal_execute(VimState *state, int key)
       break;
 
     case K_EVENT:
-      // We cannot let an event free the terminal yet. It is still needed.
-      s->term->refcount++;
       multiqueue_process_events(main_loop.events);
-      s->term->refcount--;
```

Comment:
I think it's not related to the missing refcount here, but caused by the fact that i deferred freeing terminal and job related data to when the buffer gets wiped, which vim doesn't like when exiting. Working on it, and will then update the PR.